README
WebService-Ustream-API-User version 0.01
========================================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

   version
   Class::Accessor::Fast
   LWP::UserAgent
   XML::Simple


COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2009 by Takeshi Kobayashi

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension WebService::Ustream::API.

0.02  2009-03-09
      Fixed typo on documents

0.01  Sat Mar  7 12:25:58 2009
	- original version; created by h2xs 1.23 with options
		-XA -n WebService::Ustream::API



lib/WebService/Ustream/API.pm
package WebService::Ustream::API;

use strict;
use warnings;
use version; our $VERSION = '0.02';

use base qw(Class::Accessor::Fast);

__PACKAGE__->mk_accessors(qw(key ua));

use Carp;
use WebService::Ustream::API::User;
use WebService::Ustream::API::Stream;

sub user {
	my $self = shift;

	$self->{_user} ||= WebService::Ustream::API::User->new($self);
	return $self->{_user};
}

sub stream {
	my $self = shift;
	
	$self->{_stream} ||= WebService::Ustream::API::Stream->new($self);
	return $self->{_stream};
}

1;

__END__

=head1 NAME

WebService::Ustream::API - Perl interface to Ustream

=head1 SYNOPSIS

  use WebService::Ustream::API;
  
  my $ust = WebService::Ustream::API->new( { key => YOUR_KEY } );

  # Get user information via Ustream API
  my $ret = $ust->user->info('koba206');
 
  # Get all the channels belonging to a user
  my $ret = $ust->user->list_channels('koba206');

  # Get all the videos belonging to a user
  my $ret = $ust->user->list_videos('spacevidcast');

  # Get all the comments for the user profile belonging to a user
  my $ret = $ust->user->comments('koba206'); 

  # Get information about most recent show that started broadcasting
  my $ret = $ust->stream->recent();

  # Get information about currently running show having the most viewers
  my $ret = $ust->stream->most_viewers();

  # Get information about currently on-air show that is picked randomly from all on-air broadcasts
  my $ret = $ust->stream->random();

  # Get all live shows that were created less than 1 hour old
  my $ret = $ust->stream->all_new();

=head1 DESCRIPTION

This is a Perl interface to Ustream API.
See Developer Page L<http://developer.ustream.tv/> for details.

=head1 SEE ALSO

L<URI::Fetch>
http://developer.ustream.tv/apps/register

L<URL::Fetch>
http://developer.ustream.tv/docs

=head1 AUTHOR

Takeshi Kobayashi, E<lt>tkobayashi@gmail.com<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Takeshi Kobayashi

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.


=cut


lib/WebService/Ustream/API/Stream.pm
package WebService::Ustream::API::Stream;

use strict;
use warnings;
use Carp;
use LWP::UserAgent;
use XML::Simple;

use version; our $VERSION = '0.02';

use base qw(Class::Accessor::Fast);

__PACKAGE__->mk_accessors(qw(key ua));

sub new {
        my($class, @args) = @_;

	my $self = $class->SUPER::new(@args);
	if(!$self->key){
		croak 'key is required';
	}
	if(!$self->ua){
		$self->ua( LWP::UserAgent->new );
	}
	return $self;
}

sub recent {
        my($self) = @_;

	my $uri = "http://api.ustream.tv/xml/stream/all/getRecent?key=".$self->key;

	my $result = $self->_retrieve($uri);;
        return $result;
}

sub most_viewers {
	my($self) = @_;

	my $uri = "http://api.ustream.tv/xml/stream/all/getMostViewers?key=".$self->key;
        
	my $result = $self->_retrieve($uri);
	return $result;
}

sub random {
	my($self) = @_;

        my $uri = "http://api.ustream.tv/xml/stream/all/getRandom?key=".$self->key;

        my $result = $self->_retrieve($uri);
        return $result;
}

sub all_new {
	my($self) = @_;

        my $uri = "http://api.ustream.tv/xml/stream/all/getAllNew?key=".$self->key;

	my $result = $self->_retrieve($uri);
        return $result;
}

sub _retrieve {
	my($self,$uri) = @_;

	my $res = $self->ua->get($uri);
        if( !$res->is_success ) {
                carp $res->status_line;
                return;
        }

        my $content;
        eval{$content = XMLin($res->content)};
        croak('Failed reading user information : ' . $@) if $@;
        if( !$content ) {
                carp 'invalid XML';
                return;
        }

        if( index($content->{error},'ERR') >= 0 ) {
                carp(
                        sprintf "error: %s\ndescription: %s",
                        $content->{error},
                        $content->{msg}
                );
                return;
        }
	return $content->{results};
};

1;
__END__

=head1 NAME

WebService::Ustream::API::Stream - Perl interface to Ustream Stream API Service

=head1 SYNOPSIS

  use WebService::Ustream::API::Stream;
  
  $ust = WebService::Ustream::API::Stream->new( { key => YOUR_KEY } );
  my $ret = $ust->recent(); 

  print $ret->{array}->{id};
  print $ret->{array}->{userId};
  print $ret->{array}->{userName};
  print $ret->{array}->{title};
  print $ret->{array}->{description};
  print $ret->{array}->{streamStartedAt};
  print $ret->{array}->{url};
  print $ret->{array}->{embedTag};

  my $ret = $ust->most_viewers();
  
  print $ret->{array}->{id};
  ..
  . same properties accessible as first example
  ..
  print $ret->{array}->{embedTag};

  my $ret = $ust->random();
  
  print $ret->{array}->{id};
  ..
  . same properties accessible as first example
  ..
  print $ret->{array}->{embedTag};

  my $ret = $ust->all_new();

  print $ret->{array}->{0}->{id};
  ..
  . same properties accessible as first example
  ..
  print $ret->{array}->{0}->{embedTag};


=head1 DESCRIPTION

WebService::Ustream::API::User is a simple interface to Ustream's user information.

=head1 METHODS

=item new(\%fields)

$ust = WebService::Ustream::API::User->new( { key => YOUR_KEY } );

Creates an instace of WebService::Ustream::API::User.
KEY is required when you call Ustream API.

=item get([username or userid])

  my $ret = $ust->get('koba206');
  my $ret = $ust->get('19185'); #19185 is koba206's userid 

retrieve user account information

=item list_channels([username or userid])

 my $ret = $ust->list_channels('koba206');
 
retrieve all the channels belonging to a user

=item list_videos([username or userid])

  my $ret = $ust->list_videos('spacevidcast');

retrieve all the videos belonging to the user specified

=item get_comments([username or userid])

  my $ret = $ust->get_comments('koba206');

returns any comments for the user profile specified by username or userid

=head1 SEE ALSO

L<URI::Fetch>
http://developer.ustream.tv/

=head1 AUTHOR

Takeshi Kobayashi, E<lt>koba206@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Takeshi Kobayashi

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.


=cut


lib/WebService/Ustream/API/User.pm
package WebService::Ustream::API::User;

use strict;
use warnings;
use Carp;
use LWP::UserAgent;
use XML::Simple;

use version; our $VERSION = '0.02';

use base qw(Class::Accessor::Fast);

__PACKAGE__->mk_accessors(qw(key ua));

sub new {
        my($class, @args) = @_;

	my $self = $class->SUPER::new(@args);
	if(!$self->key){
		croak 'key is required';
	}
	if(!$self->ua){
		$self->ua( LWP::UserAgent->new );
	}
	return $self;
}

sub info {
        my($self, $user) = @_;
        croak('user is required') unless $user;

	my $uri = "http://api.ustream.tv/xml/user/$user/getInfo?key=".$self->key;

	my $result = $self->_retrieve($uri);;
        return $result;
}

sub list_channels {
	my($self, $user) = @_;
        croak('user is required') unless $user;

	my $uri = "http://api.ustream.tv/xml/user/$user/listAllChannels?key=".$self->key;
        
	my $result = $self->_retrieve($uri);
        
	return $result;
}

sub list_videos {
	my($self, $user) = @_;
        croak('user is required') unless $user;

        my $uri = "http://api.ustream.tv/xml/user/$user/listAllVideos?key=".$self->key;

        my $result = $self->_retrieve($uri);

        return $result;
}

sub comments {
	my($self, $user) = @_;
        croak('user is required') unless $user;

        my $uri = "http://api.ustream.tv/xml/user/$user/getComments?key=".$self->key;

	my $result = $self->_retrieve($uri);

        return $result;
}

sub _retrieve {
	my($self,$uri) = @_;

	my $res = $self->ua->get($uri);
        if( !$res->is_success ) {
                carp $res->status_line;
                return;
        }

        my $content;
        eval{$content = XMLin($res->content)};
        croak('Failed reading user information : ' . $@) if $@;
        if( !$content ) {
                carp 'invalid XML';
                return;
        }

        if( index($content->{error},'ERR') >= 0 ) {
                carp(
                        sprintf "error: %s\ndescription: %s",
                        $content->{error},
                        $content->{msg}
                );
                return;
        }
	return $content->{results};
};

1;
__END__

=head1 NAME

WebService::Ustream::API::User - Perl interface to Ustream User API Service

=head1 SYNOPSIS

  use WebService::Ustream::API::User;
  
  $ust = WebService::Ustream::API::User->new( { key => YOUR_KEY } );
  my $ret = $ust->info('koba206'); 

  print $ret->{id};
  print $ret->{registereAt};
  print $ret->{gender};
  print $ret->{url};
  print $ret->{website};
  print $ret->{about};
  print $ret->{rating};
  print $ret->{numberOf}->{comments};
  print $ret->{numberOf}->{friends};

  my $ret = $ust->list_channels('spacevidcast');
  
  print $ret->{array}->{0}->{id};
  print $ret->{array}->{0}->{title};
  print $ret->{array}->{0}->{urlTitleName};
  print $ret->{array}->{0}->{description};
  print $ret->{array}->{0}->{totalViews};
  print $ret->{array}->{0}->{status};
  print $ret->{array}->{0}->{url};
  print $ret->{array}->{0}->{embedTag};

  my $ret = $ust->list_videos('spacevidcast');
  
  print $ret->{array}->{id};
  print $ret->{array}->{title};
  print $ret->{array}->{lengthInSecond};
  print $ret->{array}->{url};
  print $ret->{array}->{embedTag};

=head1 DESCRIPTION

WebService::Ustream::API::User is a simple interface to Ustream's user information.

=head1 METHODS

=item new(\%fields)

$ust = WebService::Ustream::API::User->new( { key => YOUR_KEY } );

Creates an instace of WebService::Ustream::API::User.
KEY is required when you call Ustream API.

=item info([username or userid])

  my $ret = $ust->info('koba206');
  my $ret = $ust->info('19185'); #19185 is koba206's userid 

retrieve user account information

=item list_channels([username or userid])

 my $ret = $ust->list_channels('koba206');
 
retrieve all the channels belonging to a user

=item list_videos([username or userid])

  my $ret = $ust->list_videos('spacevidcast');

retrieve all the videos belonging to the user specified

=item comments([username or userid])

  my $ret = $ust->comments('koba206');

returns any comments for the user profile specified by username or userid

=head1 SEE ALSO

L<URI::Fetch>
http://developer.ustream.tv/

=head1 AUTHOR

Takeshi Kobayashi, E<lt>koba206@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Takeshi Kobayashi

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.


=cut
